### ISSUE TITLE
[API Proposal]: Add WriteValue(Guid/TimeSpan) in System.Xml.XmlWriter

### ISSUE THREAD

**Issue by TrayanZapryanov** [dotnet/runtime#75515]
### Background and motivation

Writing Guids and TimeSpans in xml is quire normal practice these days. We should allow users to use it instead of casting them to sring and then using WriteValue(string) method or using WriteValue(object).

### API Proposal

```csharp
namespace System.Xml;

public class XmlWriter
{
    public virtual void WriteValue(Guid value);
    public virtual void WriteValue(TimeSpan value);
}
```

### API Usage

```csharp
XmlWriter writer;
writer.WriteValue(Guid.NewGuid());
writer.WriteValue(TimeSpan.FromHours(1));
```

### Alternative Designs

We can make these methods **internal** and use them only in System.Private.Xml project.

### Risks

none

**Comment by eiriktsarpalis** [comment]
Both `TimeSpan` and `Guid` expose `TryFormat` methods that can be used to format into a span destination. A ReadOnlySpan overload should help unlock allocation-free formatting for arbitrary types.

**Comment by drieseng** [comment]
Personally, I see good use for these proposed overloads (and I'd even add overloads for **DateOnly** and **TimeOnly**).

**Comment by eiriktsarpalis** [comment]
I have no strong opinion on adding `TimeSpan` or `Guid` overloads. For parity, the API proposal should contain equivalent overloads in `XmlReader`, the proposal needs updating to include them as well.

@krwq do you think they're worth adding?

**Comment by TahirAhmadov** [comment]
I would just add that the format of `TimeSpan`, and especially `DateOnly` and even perhaps `DateTime`, has to be carefully chosen to ensure it works across all possible values. I'm thinking it has to be "cultureless" to work.

**Comment by TrayanZapryanov** [comment]
Main benefit will be in writing Guids in XmlWriter as we will save Guid.ToString() allocation.
If you like I can remove TimeSpan as I do not expect too many TimeSpans serializations.
I've added it as it is tricky and current implementation uses XsdDuration.ToString().

**Comment by eiriktsarpalis** [comment]
@TrayanZapryanov would it be possible to update the API proposal in the OP to include `XmlReader` equivalents?

**Comment by TrayanZapryanov** [comment]
@eiriktsarpalis Done.

**Comment by eiriktsarpalis** [comment]
Thanks. I'm deferring to @krwq on whether to mark this as `api-ready-for-review` as he is the subject matter expert of this codebase.

**Comment by TrayanZapryanov** [comment]
Small risk that I've found right now:
Imagine code which is using XmlWriter.WriteValue(Guid.NewGuid()) - this will fallback to XmlWriter.WriteValue(object) if not overridden. Current implementation of it is using :
``` cs
// Writes out the specified value.
  public virtual void WriteValue(object value)
  {
      ArgumentNullException.ThrowIfNull(value);

      WriteString(XmlUntypedConverter.Untyped.ToString(value, null));
  }
```
and as **XmlUntypedConverter.Untyped.ToString(value, null)** is not supporting Guid - it will throw exception.
With proposed change - this will be accepted or in the base class we need to throw same exception.
I would prefer to accept and fallback to something like this :
```cs
// Writes out the specified value.
  public virtual void WriteValue(Guid value)
  {
      WriteString(XmlConvert.ToString(value));
  }
```

**Comment by krwq** [comment]
I'm curious - do we actually care about XmlReader APIs or we just want XmlSerializer work with those? Are you actually using reader/writer directly?

**Comment by TrayanZapryanov** [comment]
@krwq Nope, I am not using them. Proposal is just to use them when serializing primitive types in XmlSerializer.

**Comment by krwq** [comment]
I think we should start with trying to handle that in XmlSerializer directly and only add APIs to XmlReader/XmlWriter if there is clear benefit

**Comment by TrayanZapryanov** [comment]
@krwq You are right. I've added new [performance issue](https://github.com/dotnet/runtime/issues/76434).
Should I close this proposal or you see some benefit in implementing it ?

**Comment by krwq** [comment]
I'm not the owner of serialization part of the XML but I'd suggest running profiler with some real-life like payload and seeing if making a fix makes any difference. If there is clear difference then you already got prototype 😄 if change is relatively simple might be still worth to send a patch - I'd personally only do it if my particular scenario I'm trying to improve requires it (i.e. I have app which I'm trying to improve and profiler tells me this has huge impact on it or I'm particularly focusing on allocations/perf in XML)
